### PR TITLE
Document custom hooks

### PR DESCRIPTION
**Pull request by andydavison**
- Adds mdx documentation of the React SDK hooks (useSession, useDataset, useThing) within storybook
- Add folder structure within stories folder to sort stories by group (authentication, components, providers, hooks), matching the categorisation already shown in storybook

### CODE REVIEW

**Comment by andydavison** [comment]
The types of e.g. Session, Dataset, Thing, login options etc. are dependent on upstream packages, so I don't know how much detail we can/should go into describing those in these docs? At present they aren't detailed here.

**Comment by nicolasmondada** [comment]
> The types of e.g. Session, Dataset, Thing, login options etc. are dependent on upstream packages, so I don't know how much detail we can/should go into describing those in these docs? At present they aren't detailed here.

They are indeed dependent of our upstream packages. We should link to docs.inrupt.com where possible and if there is no description to link to please open a ticket on DOCS

**Comment by ajacksified** [comment]
I'd suggest reordering useSession to the top, and useDataset and useThing afterwards.

File: stories/hooks/useDataset.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+Should be located within a `SessionProvider` if using to fetch.
+
+```
+const { thing, error } = useDataset(datasetIri, options)
+```
+
+## Parameters
+
+- `datasetIri`: (optional) a dataset iri to fetch from
```
Comment (nicolasmondada) [inline stories/hooks/useDataset.stories.mdx:17]:
```suggestion
- `datasetIri`: (optional) a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) to fetch from
```

File: stories/hooks/useThing.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+Should be located within a `SessionProvider` if using to fetch.
+
+```
+const { thing, error } = useThing(datasetIri, thingIri, options)
+```
+
+## Parameters
+
+- `datasetIri`: (optional) a dataset iri to fetch the Thing from, if both `datasetIri` and `thingIri` are supplied. Uses the Dataset from context if omitted.
+- `thingIri`: (optional) a thing iri to fetch the Thing from. Uses the Thing from context if omitted.
```
Comment (nicolasmondada) [inline stories/hooks/useThing.stories.mdx:18]:
```suggestion
- `thingIri`: (optional) a [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) to fetch the [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from. Uses the [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from context if omitted.
```

File: stories/hooks/useThing.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+- `options`: (optional) additional options to pass to the dataset request, if `datasetIri` is supplied
```
Comment (nicolasmondada) [inline stories/hooks/useThing.stories.mdx:19]: what are these options?

**Review by nicolasmondada — CHANGES_REQUESTED** [review]

File: stories/hooks/useDataset.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Meta } from "@storybook/addon-docs/blocks";
+
+<Meta title="Hooks/useDataset" />
+
+# useDataset
+
+A hook to access the nearest Dataset from context if no arguments are supplied, or to request a Dataset from given IRI.
+
+Should be located within a `SessionProvider` if using to fetch.
```
[the remark this answers is not in the document]
Comment (andydavison) [reply stories/hooks/useDataset.stories.mdx]: after some research, seems so! Have added

File: stories/hooks/useThing.stories.mdx (hunk 3 above)
Comment (andydavison) [reply stories/hooks/useThing.stories.mdx:20]: the options here: https://docs.inrupt.com/developer-tools/api/javascript/solid-client/modules/_resource_soliddataset_.html#getsoliddataset

The only one actually documented there is `fetch`, which is the one useDataset/useThing deal with automatically

Comment (andydavison) [reply stories/hooks/useThing.stories.mdx]: @nicolasmondada I've just added links to those getsoliddataset docs

**Comment by andydavison** [comment]
> I'd suggest reordering useSession to the top, and useDataset and useThing afterwards.

This is now updated ✅

**Review by ajacksified — APPROVED** [review]

File: stories/hooks/useDataset.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { Meta, Source } from "@storybook/addon-docs/blocks";
+
+<Meta title="Hooks/useDataset" />
+
+# useDataset
+
+A hook to access the nearest Dataset from context if no arguments are supplied, or to request a Dataset from given IRI.
```
Comment (nicolasmondada) [inline stories/hooks/useDataset.stories.mdx:7]:
```suggestion
A hook to access the nearest [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) from context if no arguments are supplied, or to request a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) from given [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI).
```

File: stories/hooks/useDataset.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+## Parameters
+
+- `datasetIri`: (optional) a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) to fetch from
+- `options`: (optional) additional options to pass to the [dataset request](https://docs.inrupt.com/developer-tools/api/javascript/solid-client/modules/_resource_soliddataset_.html#getsoliddataset), if `datasetIri` is supplied. Unless an alternative id provided, `options.fetch` is included using the current session automatically.
+
+## Return Values
+
+- `dataset`: the Dataset retrieved from context or request using given iri (undefined if loading/missing)
```
Comment (nicolasmondada) [inline stories/hooks/useDataset.stories.mdx:22]:
```suggestion
- `dataset`: the [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) retrieved from context or request using given [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) (undefined if loading/missing)
```

File: stories/hooks/useDataset.stories.mdx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+- `error`: the error thrown by the request (or undefined)
+
+## Example Usage
+
+This example uses the datasetIri value to fetch a dataset, and pass to a child component when loaded.
```
Comment (nicolasmondada) [inline stories/hooks/useDataset.stories.mdx:27]:
```suggestion
This example uses the datasetIri value to fetch a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset), and pass to a child component when loaded.
```

File: stories/hooks/useThing.stories.mdx [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import { Meta, Source } from "@storybook/addon-docs/blocks";
+import { useThing } from "../../src";
+
+<Meta title="Hooks/useThing" />
+
+# useThing
+
+A hook to access the nearest [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from context if no arguments are supplied, or to request a [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from a given IRI.
```
Comment (nicolasmondada) [inline stories/hooks/useThing.stories.mdx:8]:
```suggestion
A hook to access the nearest [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from context if no arguments are supplied, or to request a [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from a given [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI).
```

File: stories/hooks/useThing.stories.mdx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+Should be located within a [`SessionProvider`](/?path=/docs/authentication-session-provider) if using to fetch.
+
+```
+const { thing, error } = useThing(datasetIri, thingIri, options)
+```
+
+## Parameters
+
+- `datasetIri`: (optional) a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) to fetch the [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from, if both `datasetIri` and `thingIri` are supplied. Uses the Dataset from context if omitted.
```
Comment (nicolasmondada) [inline stories/hooks/useThing.stories.mdx:18]:
```suggestion
- `datasetIri`: (optional) a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) to fetch the [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) from, if both `datasetIri` and `thingIri` are supplied. Uses the [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) from context if omitted.
```

File: stories/hooks/useThing.stories.mdx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+- `options`: (optional) additional options to pass to the [dataset request](https://docs.inrupt.com/developer-tools/api/javascript/solid-client/modules/_resource_soliddataset_.html#getsoliddataset), if `datasetIri` is supplied. Unless an alternative id provided, `options.fetch` is included using the current session automatically.
+
+## Return Values
+
+- `thing`: the [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) retrieved from context or requests using given [IRI](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-IRI) (undefined if loading/missing)
+- `error`: the error thrown by the request (or undefined)
+
+## Example Usage
+
+This example uses the datasetIri and thingIri values to fetch a thing, and pass to a child component when loaded.
```
Comment (nicolasmondada) [inline stories/hooks/useThing.stories.mdx:29]:
```suggestion
This example uses the datasetIri and thingIri values to fetch a [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing), and pass to a child component when loaded.
```

**Review by nicolasmondada — CHANGES_REQUESTED** [review]

**Review by nicolasmondada — APPROVED** [review]

**Pull request merged**